parachains/common/src/xcm_config.rs

```diff
@@ -156,6 +161,59 @@ impl<
 	}
 }
 
+/// Allows execution from `origin` if it is contained in `AllowedOrigin`
+/// and if it is just a straight `Transact` which contains `AllowedCall`.
+pub struct AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>(
+	sp_std::marker::PhantomData<(RuntimeCall, AllowedCall, AllowedOrigin)>,
+);
+impl<
+		RuntimeCall: Decode,
+		AllowedCall: Contains<RuntimeCall>,
+		AllowedOrigin: Contains<MultiLocation>,
+	> ShouldExecute for AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		max_weight: Weight,
+		_properties: &mut xcm_executor::traits::Properties,
+	) -> Result<(), ProcessMessageError> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowUnpaidTransactFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, properties: {:?}",
+			origin, instructions, max_weight, _properties,
+		);
+
+		// we only allow from configured origins
+		ensure!(AllowedOrigin::contains(origin), ProcessMessageError::Unsupported);
+
+		// we expect an XCM program with single `Transact` call
+		instructions
+			.matcher()
+			.assert_remaining_insts(1)?
+			.match_next_inst(|inst| match inst {
+				Transact { origin_kind: OriginKind::Xcm, call: encoded_call, .. } => {
+					// this is a hack - don't know if there's a way to do that properly
+					// or else we can simply allow all calls
+					let mut decoded_call = DoubleEncoded::<RuntimeCall>::from(encoded_call.clone());
+					ensure!(
+						AllowedCall::contains(
+							decoded_call
+								.ensure_decoded()
+								.map_err(|_| ProcessMessageError::BadFormat)?
+						),
+						ProcessMessageError::BadFormat,
+					);
```

Comment:
Unless I miss something, the `AllowUnpaidTransactsFrom` is configured with `AllowedCall` set to calls of `pallet-xcm-bridge-hub-router`:
```rust
	impl Contains<RuntimeCall> for ToPolkadotXcmRouter {
		fn contains(call: &RuntimeCall) -> bool {
			matches!(
				call,
				RuntimeCall::ToPolkadotXcmRouter(
					pallet_xcm_bridge_hub_router::Call::report_bridge_status { .. }
				)
			)
		}
	}
...
bridging::AllowUnpaidStatusReportsFromSiblingBridgeHub<ToPolkadotXcmRouter>,
...
mod bridging {
...
	/// Barrier for `pallet_xcm_bridge_hub_router::Pallet` to receive congestion statuses from
	/// sibling BridgeHub.
	pub type AllowUnpaidStatusReportsFromSiblingBridgeHub<XcmBridgeHubRouter> =
		AllowUnpaidTransactsFrom<RuntimeCall, XcmBridgeHubRouter, Equals<BridgeHubKusama>>;
...
}
```

So it'll return `contains` only for the call we're interested in. And for better understanding it is better to read `.contains()` here as `.eq()` - because it doesn't iterate through calls in the `decoded_call` - it just compares the `decoded_call` with the `AllowedCall`s. And as I mentioned, our `AllowedCall` is the `report_bridge_status` - so all `utility.batchAll` and so on will be rejected.

So imo all should be fine. But actually this particular struct is maybe an overkill (I was experimenting with filters and as you see it works with some hacks - e.g. decoding) - I think we may allow unpaid calls from BH on AH for v1. But I'm fine with the current version too